pkg/providers/snow/snow.go

```diff
@@ -221,7 +223,25 @@ func (p *SnowProvider) MachineConfigs(clusterSpec *cluster.Spec) []providers.Mac
 	return configs
 }
 
+// ValidateNewSpec validates the immutability of snow machine config by comparing it with the existing one in cluster.
 func (p *SnowProvider) ValidateNewSpec(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+	for _, mc := range clusterSpec.SnowMachineConfigs {
+		oldMc := &v1alpha1.SnowMachineConfig{}
+		err := p.kubeUnAuthClient.Get(ctx, mc.GetName(), mc.GetNamespace(), cluster.KubeconfigFile, oldMc)
+
+		// if machine config object does not exist in cluster, it means user defines new machine config. Skip comparison.
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		if oldMc.Spec.SshKeyName != mc.Spec.SshKeyName {
```

Comment:
Need to double check with snow team since technically ssh key name update should be durable since new machine will be created.